### PR TITLE
Add Program Management modal popup

### PR DESCRIPTION
**Pull request by toddburnside**
Adds the ability to select/add/delete/edit programs from the menu, or in the case where a program is not specified in the URL. In the latter case, if the user only has one program, it will select that program without showing the popup. This does not yet handle the case of an invalid ID in the URL. 

Also, at this point editing the program name does not persist because the API version in development and staging does not support it. It is supported in API master, but upgrading Explore to use it is not straight forward since API master also contains change to the sequences. I think it will be better to wait for Raul's changes.

I opted for the ability to show deleted programs over undo/redo because I think it could be important for a user to be able to undelete a program even if they do something like reload the page. Undo/redo could be added as well if it becomes regarded as important for name editing.

image

### CODE REVIEW

File: common/src/main/scala/explore/model/reusability.scala [hunk 1]
```diff
@@ -70,4 +71,6 @@ object reusability {
   implicit val catalogTargetResultReuse: Reusability[CatalogTargetResult]          = Reusability.derive
   implicit val scienceConfigurationnResultReuse: Reusability[ScienceConfiguration] =
     Reusability.byEq
+
+  implicit val existenceReuse: Reusability[Existence] = Reusability.derive
```
Comment (toddburnside) [inline common/src/main/scala/explore/model/reusability.scala:75]: I'm not sure if there is a better place for this?

Comment (cquiroz) [reply common/src/main/scala/explore/model/reusability.scala:75]: This is probably the right place

File: explore/src/main/scala/explore/programs/ProgramsPopup.scala [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+          ProgramTable(
+            props.currentProgramId,
+            programs,
+            showDeleted.get,
+            selectProgram = Reuse.currying(isRequired, isOpen, props.undoStacks).in(selectProgram _)
+          ),
+          <.div(
+            Button(
+              clazz = ExploreStyles.ProgramAdd,
+              // size = Small,
```
Comment (cquiroz) [inline explore/src/main/scala/explore/programs/ProgramsPopup.scala:128]: Commentted code

File: explore/src/main/scala/explore/programs/ProgramsPopup.scala [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            )
+          )
+        )
+      )
+    )
+  }
+
+  protected val component = ScalaFnComponent
+    .withHooks[Props]
+    .useStateViewWithReuse(false) // isOpen
```
Comment (cquiroz) [inline explore/src/main/scala/explore/programs/ProgramsPopup.scala:148]: I hope when we have opaque types we could make these so it is clearer which one is doing what, but that's unrelated to this PR

Comment (toddburnside) [reply explore/src/main/scala/explore/programs/ProgramsPopup.scala:148]: Right. There are a lot of Booleans. ☹️

**Comment by cquiroz** [comment]
Yeah, the eraser looks nice

**Pull request merged**